DoS protection: Weighted random drop of txs if mempool full

Eirik0 opened the pull request.

daira commented on qa/rpc-tests/mempool_limit.py:
```diff
@@ -0,0 +1,90 @@
+    def setup_nodes(self):
+        args = [
+            '-nuparams=5ba81b19:1', # Overwinter
+            '-nuparams=76b809bb:1', # Sapling
```
These `-nuparams` args are no longer needed after #4060.

daira commented on qa/rpc-tests/mempool_limit.py:
```diff
@@ -0,0 +1,90 @@
+
+        print("Adding one more transaction...")
+        self.nodes[3].sendtoaddress(taddr3, 9.999)
+        # The mempools are no longer guarenteed to be in a consistent state, so we cannot sync
```
```suggestion
        # The mempools are no longer guaranteed to be in a consistent state, so we cannot sync
```

daira commented on qa/rpc-tests/mempool_limit.py:
```diff
@@ -0,0 +1,90 @@
+            assert_equal(2, len(mempool), "node {}".format(i))
+
+        # self.nodes[0].generate(1)
+        # self.sync_all()
```
Delete these lines.

daira commented on qa/rpc-tests/mempool_limit.py:
```diff
@@ -0,0 +1,90 @@
+        # self.nodes[0].generate(1)
+        # self.sync_all()
+
+        print("Success")
```
Also delete this.

daira commented on src/gtest/test_mempoollimit.cpp:
```diff
@@ -0,0 +1,140 @@
+const uint256 TX_ID2 = ArithToUint256(2);
+const uint256 TX_ID3 = ArithToUint256(3);
+
+TEST(MempoolLimitTests, RecentlyEvictedList_AddWrapsAfterMaxSize)
```
https://github.com/google/googletest/blob/master/googletest/docs/faq.md#why-should-test-suite-names-and-test-names-not-contain-underscore

daira commented on src/gtest/test_mempoollimit.cpp:
```diff
@@ -0,0 +1,140 @@
+    recentlyEvicted.add(TX_ID1);
+    recentlyEvicted.add(TX_ID2);
+    recentlyEvicted.add(TX_ID3);
+    // tx 1 should be overwritten by tx 3 due maxSize 2
```
```suggestion
    // tx 1 should be overwritten by tx 3 due to maxSize 2
```

daira commented on src/gtest/test_mempoollimit.cpp:
```diff
@@ -0,0 +1,140 @@
+    // tx 1 should be overwritten by tx 3 due maxSize 2
+    EXPECT_FALSE(recentlyEvicted.contains(TX_ID1));
+    EXPECT_TRUE(recentlyEvicted.contains(TX_ID2));
+    EXPECT_TRUE(recentlyEvicted.contains(TX_ID3));
```
Perhaps test that `TX_ID2` is the next to be evicted. (I can imagine a buggy queue that would pass the test as it stands.)

Eirik0 replied:
This has now been added this as a separate test.

daira commented on src/init.cpp:
```diff
@@ -401,6 +401,8 @@ std::string HelpMessage(HelpMessageMode mode)
     strUsage += HelpMessageOpt("-whitelist=<netmask>", _("Whitelist peers connecting from the given netmask or IP address. Can be specified multiple times.") +
         " " + _("Whitelisted peers cannot be DoS banned and their transactions are always relayed, even if they are already in the mempool, useful e.g. for a gateway"));
+    strUsage += HelpMessageOpt("-mempooltotalcostlimit=<n>", _("An upper bound on the maximum size in bytes of all txs in the mempool. (default: 80000000)"));
+    strUsage += HelpMessageOpt("-mempoolevictionmemoryminutes=<n>", _("The number of minutes before reallowing rejected transactions from reentering the mempool. (default: 60)"));
```
```suggestion
    strUsage += HelpMessageOpt("-mempoolevictionmemoryminutes=<n>", _("The number of minutes before reallowing rejected transactions to reenter the mempool. (default: 60)"));
```

Eirik0 commented:
This is the implementation for https://github.com/zcash/zips/pull/275

str4d commented on src/main.cpp:
```diff
@@ -1598,17 +1602,29 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
+            boost::optional<uint256> maybeDropTxId;
+            list<CTransaction> removed;
+            while ((maybeDropTxId = pool.ensureSizeLimit()).is_initialized()) {
+                pool.remove(pool.mapTx.find(maybeDropTxId.get())->GetTx(), removed, true);
```
Why is the looping and removal not just part of `CTxMemPool::ensureSizeLimit`? As it is, the call does _not_ ensure the size limit is met, because the caller can choose to ignore the return value.

str4d commented on src/txmempool.cpp:
```diff
@@ -752,3 +761,27 @@ size_t CTxMemPool::DynamicMemoryUsage() const {
+    assert(!recentlyEvicted);
+    assert(!weightedTxList);
+    recentlyEvicted = new RecentlyEvictedList(evictionMemorySeconds);
+    weightedTxList = new WeightedTransactionList(totalCostLimit);
```
If we are storing these as pointers, we need to explicitly free them in `CTxMemPool::~CTxMemPool`.

str4d commented on src/txmempool.cpp:
```diff
@@ -752,3 +761,27 @@ size_t CTxMemPool::DynamicMemoryUsage() const {
+    return recentlyEvicted->contains(txId);
+}
+
+boost::optional<uint256> CTxMemPool::ensureSizeLimit() {
```
Missing `LOCK(cs);` (for `weightedTxList`).

Eirik0 commented:
I rewrote the representation of the collection of transactions/weights because I was concerned about performance when removing transactions from the mempool. In the first iteration we rebuild the complete list of weighted transactions each time a transaction was removed from the mempool. Now, we store the transactions in a tree and are able to add and remove transactions from the collection in logarithmic time.

I also did some more cleanup and rebased on to master now that #4060 has been merged.

zebambam commented on src/init.cpp:
```diff
@@ -383,6 +384,8 @@ std::string HelpMessage(HelpMessageMode mode)
     strUsage += HelpMessageOpt("-maxreceivebuffer=<n>", strprintf(_("Maximum per-connection receive buffer, <n>*1000 bytes (default: %u)"), 5000));
     strUsage += HelpMessageOpt("-maxsendbuffer=<n>", strprintf(_("Maximum per-connection send buffer, <n>*1000 bytes (default: %u)"), 1000));
+    strUsage += HelpMessageOpt("-mempoolevictionmemoryminutes=<n>", strprintf(_("The number of minutes before allowing rejected transactions to re-enter the mempool. (default: %u)"), DEFAULT_MEMPOOL_EVICTION_MEMORY_MINUTES));
+    strUsage += HelpMessageOpt("-mempooltotalcostlimit=<n>",strprintf(_("An upper bound on the maximum size in bytes of all transactions in the mempool. (default: %s)"), DEFAULT_MEMPOOL_TOTAL_WEIGHT_LIMIT));
```
still called costlimit even though it's now a size limit

zebambam commented on src/mempoollimit.cpp:
```diff
@@ -0,0 +1,159 @@
+const CAmount DEFAULT_FEE = 10000;
+const TxWeight ZERO_WEIGHT = TxWeight(0, 0);
+
+void RecentlyEvictedList::pruneList()
```
I don't understand how this is locked.

zebambam commented on src/mempoollimit.cpp:
```diff
@@ -0,0 +1,159 @@
+    pruneList();
+    if (txIdsAndTimes[txIdsAndTimesIndex].is_initialized()) {
+        auto txIdAndTime = txIdsAndTimes[txIdsAndTimesIndex];
+        txIdSet.erase(txIdAndTime.get().first);
```
So.. we just overwrite anything existing in there? This seems like someone could fill up the RecentlyEvictedList with permutations of (malleable) transactions and we still end up in DoS. What have I missed?

ebfull replied:
Note that the `RecentlyEvictedList` is not intended to mitigate intentional DoS.

Eirik0 replied:
We could make the size of the recently evicted list configurable, but I am not sure about how beneficial that would be.

zebambam commented on src/mempoollimit.h:
```diff
@@ -0,0 +1,95 @@
+#include "uint256.h"
+
+const size_t DEFAULT_MEMPOOL_TOTAL_WEIGHT_LIMIT = 80000000;
+const int64_t DEFAULT_MEMPOOL_EVICTION_MEMORY_MINUTES = 60;
```
I don't understand the considerations that went into this figure.

daira replied:
They're described in the draft ZIP.

zebambam commented on src/mempoollimit.h:
```diff
@@ -0,0 +1,95 @@
+#include "primitives/transaction.h"
+#include "uint256.h"
+
+const size_t DEFAULT_MEMPOOL_TOTAL_WEIGHT_LIMIT = 80000000;
```
Do you mean weight here or size? I don't understand the considerations that went into this figure.

LarryRuane replied:
@zebambam, https://github.com/zcash/zips/pull/275/files#diff-9be987255d87572395564cded9f3c860R94 (and similarly with some other constants). It is weight.

zebambam commented on src/mempoollimit.h:
```diff
@@ -0,0 +1,95 @@
+const size_t DEFAULT_MEMPOOL_TOTAL_WEIGHT_LIMIT = 80000000;
+const int64_t DEFAULT_MEMPOOL_EVICTION_MEMORY_MINUTES = 60;
+
+const size_t RECENTLY_EVICTED_SIZE = 10000;
```
I don't understand the considerations that went into this figure.

Eirik0 replied:
This is a mistake, I must have misread/misremembered. Thanks for catching that.

ebfull commented on src/mempoollimit.cpp:
```diff
@@ -0,0 +1,159 @@
+void WeightedTxTree::remove(const uint256& txId)
+{
+    if (txIdToIndexMap.find(txId) == txIdToIndexMap.end()) {
+        return;
```
This should never happen anyway, right?

Eirik0 replied:
In testing I found that this does actually happen - somewhere remove is being called on transactions that have already been removed. I did not research where or why, but as it stands this is necessary.

daira replied:
Please file a ticket to investigate this.

Eirik0 replied:
I created https://github.com/zcash/zcash/issues/4159, but I question how much time it is worth spending on this.

ebfull commented on src/mempoollimit.cpp:
```diff
@@ -0,0 +1,159 @@
+
+    size_t removeIndex = txIdToIndexMap[txId];
+
+    TxWeight lastChildWeight = txIdAndWeights[--size].txWeight;
```
`size` should be decremented explicitly, not as a side-effect of this line.

Eirik0 replied:
I personally like inlining `++` and `--` where I can, but I will change this for readability.

ebfull commented on src/mempoollimit.cpp:
```diff
@@ -0,0 +1,159 @@
+    txIdAndWeights.push_back(weightedTxInfo);
+    childWeights.push_back(ZERO_WEIGHT);
+    txIdToIndexMap[weightedTxInfo.txId] = size;
+    backPropagate(size++, weightedTxInfo.txWeight);
```
`size` should be explicitly incremented on its own line.

ebfull commented on src/mempoollimit.cpp:
```diff
@@ -0,0 +1,159 @@
+WeightedTxInfo WeightedTxInfo::from(const CTransaction& tx, const CAmount& fee)
+{
+    size_t memUsage = RecursiveDynamicUsage(tx);
+    memUsage += tx.vJoinSplit.size() * JOINSPLIT_SIZE;
```
Should these instead be folded into `RecursiveDynamicUsage`?

Eirik0 replied:
Should this be fixed as part of this issue, or can it be done as follow up?

Eirik0 replied:
If we did this as follow up, we could also incorporate this in to `z_mergetoaddress` which does its own tx size estimation, but in a different way.

ebfull commented on src/mempoollimit.cpp:
```diff
@@ -0,0 +1,159 @@
+}
+
+
+// These are also defined in rpcwallet.cpp
```
They should be defined once (probably in `transaction.h`)

ebfull commented on src/mempoollimit.cpp:
```diff
@@ -0,0 +1,159 @@
+    backPropagate(size, lastChildWeight.negate());
+
+    if (removeIndex < size) {
+        TxWeight weightDelta = lastChildWeight.add(txIdAndWeights[removeIndex].txWeight.negate());
```
Couldn't this potentially cause an integer underflow?

Eirik0 replied:
Yes, the weights/costs should not be defined as unsigned. I will fix this.

LarryRuane commented on src/txmempool.h:
```diff
@@ -260,6 +263,12 @@ class CTxMemPool
+    // Returns true if a transaction has been recently evicted
+    bool IsRecentlyEvicted(const uint256& txId);
+    // Returns a txid if a transaction is evicted from the mempool
+    void EnsureSizeLimit();
```
Is this comment wrong? Does not return a txid?

Eirik0 commented on qa/rpc-tests/mempool_limit.py:
```diff
@@ -0,0 +1,103 @@
+        self.nodes[3].sendtoaddress(taddr3, 9.999)
+        # The mempools are no longer guaranteed to be in a consistent state, so we cannot sync
+        sleep(5)
+        mempool_node3 = self.nodes[i].getrawmempool()
```
This was an error. It should now be fixed.

Eirik0 commented on qa/rpc-tests/mempool_limit.py:
```diff
@@ -0,0 +1,103 @@
+
+        for i in range(0, 4):
+            mempool = self.nodes[i].getrawmempool()
+            print("Mempool for node {}: {}".format(i, mempool))
```
I left some of the prints, but changed it so it will only print the nodes mempools in the case of a that one is the wrong size.

LarryRuane commented on src/mempoollimit.h:
```diff
@@ -0,0 +1,95 @@
+    const int64_t timeToKeep;
+    // Pairs of txid and time (seconds since epoch)
+    boost::optional<std::pair<uint256, int64_t>> txIdsAndTimes[RECENTLY_EVICTED_SIZE];
+    std::set<uint256> txIdSet;
```
Couldn't both of these data structures be replaced by a `std::deque`? And the code made quite a bit simpler and more efficient? A deque is perfect for implementing a FIFO.

Eirik0 replied:
I made the change to use a `std::deque` and it did simplify the code and should reduce the memory footprint in normal circumstances.

Using an `unordered_set` rather than a `set` does not compile and I think would require making `uint256` hashable.

Eirik0 commented on src/mempoollimit.cpp:
```diff
@@ -0,0 +1,159 @@
+        return boost::none;
+    }
+    LogPrint("mempool", "Mempool cost limit exceeded (cost=%d, limit=%d)\n", totalPenaltyWeight, capacity);
+    int randomWeight = GetRand(totalPenaltyWeight);
```
This should be using the 'weight' rather than the 'lowFeePenaltyWeight' when comparing against the capacity.

Eirik0 commented on src/Makefile.am:
```diff
@@ -210,6 +210,7 @@ BITCOIN_CORE_H = \
   torcontrol.h \
   transaction_builder.h \
   txdb.h \
+  mempoollimit.h \
```
I went ahead and changed this.

LarryRuane commented on src/mempool_limit.h:
```diff
@@ -0,0 +1,95 @@
+    const int64_t timeToKeep;
+    // Pairs of txid and time (seconds since epoch)
+    std::deque<std::pair<uint256, int64_t>> txIdsAndTimes;
+    std::set<uint256> txIdSet;
```
I figured out how to do an unordered set (more efficient), if you'd like. The gtests pass with this change.
```
diff --git a/src/mempool_limit.h b/src/mempool_limit.h
index 63b24eaa1..89cf405c6 100644
--- a/src/mempool_limit.h
+++ b/src/mempool_limit.h
@@ -6,7 +6,7 @@
 #define MEMPOOLLIMIT_H
 
 #include <map>
-#include <set>
+#include <unordered_set>
 #include <vector>
 
 #include "boost/optional.hpp"
@@ -20,6 +20,13 @@ const size_t RECENTLY_EVICTED_SIZE = 40000;
 const uint64_t MIN_TX_WEIGHT = 4000;
 const uint64_t LOW_FEE_PENALTY = 16000; 
 
+class RecentlyEvictedListHasher {
+public:
+    size_t operator()(const uint256& key) const {
+        return key.GetCheapHash();
+    }
+};
+
 class RecentlyEvictedList
 {
     const size_t capacity;
@@ -27,7 +34,7 @@ class RecentlyEvictedList
     const int64_t timeToKeep;
     // Pairs of txid and time (seconds since epoch)
     std::deque<std::pair<uint256, int64_t>> txIdsAndTimes;
-    std::set<uint256> txIdSet;
+    std::unordered_set<uint256, RecentlyEvictedListHasher> txIdSet;
 
     void pruneList();
 
```

LarryRuane replied:
In case it's not obvious, this suggestion is non-blocking. We can possibly do this later, or maybe create a tech-debt PR that replaces all unnecessarily-ordered sets and maps with unordered ones.

Eirik0 replied:
Nice. Looks great! At this point I think I would be happier fixing these as part of a follow up, but if other reviewers think this should this should be included as part of this PR I am happy to do that to.

LarryRuane commented on src/mempool_limit.h:
```diff
@@ -0,0 +1,95 @@
+    void remove(const uint256& txId);
+
+    boost::optional<uint256> maybeDropRandom();
+};
```
This is a cool object, with cool algorithms, but are we sure enough that the performance will be unacceptable if we just iterate the mempool (`CTxMemPool::mapTx`) each time we need to evict a tx? At most there will be 20k transactions, and on average we'll visit half of those. Is the overhead of iterating 10k transactions small compared with the overhead of actually adding and removing a transaction? If so, then the iteration overhead wouldn't be significant.

If the iteration overhead is significant, here's another idea I just thought of: We could evict (say) 10 transactions at a time. That introduces some error, but 10 is a tiny fraction of 20k. (It's true that we could be at the weight limit with fewer transactions, but 10 would still be a small fraction.) This doesn't conform to the https://github.com/zcash/zips/pull/275, but would the deviation be acceptable?

Here's a way to evict 10 at a time with a single iteration pass. Generate a vector of 10 random numbers between 1 and the total current weight (or one less than these values). Sort the vector low-to-high, then calculate differences. So `[2, 5, 14]` (after sorting) would become `[2, 3, 9]`. Start iterating the mempool, decrementing the first value in the vector by each transaction's weight, and if it goes negative, then evict this transaction (or add it to a temporary eviction list if we can't evict it without breaking the iterator). Discard the first value in the vector (probably by just incrementing the index). Continue iterating and evicting.

A minor variation would be to not evict these 10 immediately, but only save a list of their tx IDs. Then, when adding a new tx that would exceed the weight limit, evict the next tx on the list, if it still exists. (This would not conform to the zip in that the new tx wouldn't be eligible for eviction, but perhaps that would be acceptable.) When the list is empty, iterate to create a list of 10 more.

The nice thing about this approach is we wouldn't need an entirely new data structure; the existing `mapTx` would be sufficient.

Eirik0 replied:
I think that your right in that this could be done by just iterating over `mapTx`, recalculating transaction sizes as we go until we find the transaction that we want to evict. We would also have to keep track of the the total size of the mempool as transactions are added and removed, but that shouldn't be too difficult. I like the simplicity of that solution and I wish I had thought of this when I first started working on this. In light of your suggestion, it is likely that current implementation is prematurely optimized. I haven't done any performance analysis but my intuition tells me that iterating over `mapTx` would be sufficient, but also that my current solution is more performant. That being said, the 2.1.0 release is blocked on solving this issue and I am hesitant to redo this. 🤷‍♂

zebambam commented:
On the basis that this is strictly better than what we are currently doing, according to my analysis...

zebambam commented:
utACK

daira commented on src/mempool_limit.cpp:
```diff
@@ -0,0 +1,160 @@
+
+WeightedTxInfo WeightedTxInfo::from(const CTransaction& tx, const CAmount& fee)
+{
+    size_t memUsage = RecursiveDynamicUsage(tx);
```
Shouldn't `RecursiveDynamicUsage` be changed to include the sizes of JoinSplit, Output, and Spend descriptions?

Eirik0 replied:
Yes, I will file a separate ticket for this.

Eirik0 replied:
https://github.com/zcash/zcash/issues/4158

LarryRuane commented on src/main.cpp:
```diff
@@ -1616,17 +1621,24 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
-        // insightexplorer: Add memory spent index
-        if (fSpentIndex) {
-            pool.addSpentIndex(entry, view);
+            pool.EnsureSizeLimit();
```
We may be able to simplify all this by merging the two insightexplorer calls and the call to `EnsureSizeLimit()` into `addUnchecked()`, which is called from only this one place. Then `addUnchecked` can acquire the lock and it wouldn't have to be done here. However, `addUnchecked()` is called from several places in test code, which may break if this function does these additional things. Probably best to leave as it is here for now at least. 

One small change you could make is to have the two insightexplorer functions and `EnsureSizeLimit()` assert that the lock is held, instead of (recursively) calling `LOCK(cs)`.

Eirik0 replied:
I like this suggestion, but let's do this as follow up. As a note, this may also require adding a call to `CTxMempool::SetMempoolCostLimit(...)` in the tests which call `CTxMempool::addUnchecked(...)`.

daira commented on doc/release-notes.md:
```diff
@@ -4,6 +4,20 @@ release-notes at release time)
+mempool. The maximum total cost is configurable via the parameter
+`mempooltotalcostlimit` which defaults to 80,000,000 (up to 20,000 txs). If a
+node's total mempool `cost` exceeds this limit the node will evict a random
+transaction, weighted by its cost. To prevent a node from re-accepting evicted
```
This description is no longer accurate because of the separation of cost and eviction weight. The cost is not dependent on the fee; the eviction weight is.

daira replied:
Also, the parameter names do not match the draft ZIP. Either the ZIP or the code (and this description), or both, has to change. Personally I like the `.` after `mempool` to separate it into a different namespace. snake_case vs mushedlowercase is a matter of taste.

Eirik0 replied:
How about changing this to say ... `a random transaction,`**`preferentially picking larger transactions and ones with below the standard fee.`**`To prevent`...?

daira replied:
Note that "and its fee" also needs to be deleted from the definition of cost.

Eirik0 replied:
I changed the wording a bit here for clarification.

daira commented on src/mempool_limit.h:
```diff
@@ -0,0 +1,129 @@
+    std::vector<TxWeight> childWeights;
+
+    // The following map is to simplify removal. When removing a tx, we do so by txid.
+    // This map allows look up the transactions index in the tree.
```
```suggestion
    // This map allows looking up the transaction's index in the tree.
```

daira commented on src/mempool_limit.h:
```diff
@@ -0,0 +1,129 @@
+    // This map allows look up the transactions index in the tree.
+    std::map<uint256, size_t> txIdToIndexMap;
+
+    // Returns the sum of a node and all of its childrens' TxWeights for a given index.
```
```suggestion
    // Returns the sum of a node and all of its children's TxWeights for a given index.
```

daira replied:
(Yes English is very inconsistent. http://www.everythingenglishblog.com/?p=269 )

Eirik0 commented:
@zkbot try

Eirik0 commented:
Intermittent failures.

LarryRuane commented on src/init.cpp:
```diff
@@ -384,8 +384,8 @@ std::string HelpMessage(HelpMessageMode mode)
-    strUsage += HelpMessageOpt("-mempoolevictionmemoryminutes=<n>", strprintf(_("The number of minutes before allowing rejected transactions to re-enter the mempool. (default: %u)"), DEFAULT_MEMPOOL_EVICTION_MEMORY_MINUTES));
-    strUsage += HelpMessageOpt("-mempooltotalcostlimit=<n>",strprintf(_("An upper bound on the maximum size in bytes of all transactions in the mempool. (default: %s)"), DEFAULT_MEMPOOL_TOTAL_COST_LIMIT));
+    strUsage += HelpMessageOpt("-mempool.eviction_memory_minutes=<n>", strprintf(_("The number of minutes before allowing rejected transactions to re-enter the mempool. (default: %u)"), DEFAULT_MEMPOOL_EVICTION_MEMORY_MINUTES));
+    strUsage += HelpMessageOpt("-mempool.tx_cost_limit=<n>",strprintf(_("An upper bound on the maximum size in bytes of all transactions in the mempool. (default: %s)"), DEFAULT_MEMPOOL_TOTAL_COST_LIMIT));
```
Sorry, but I think the old way is better, because, as Eirik suspected, `.` does seem to have a special meaning at least in Bitcoin:
https://github.com/bitcoin/bitcoin/blob/b6e34afe9735faf97d6be7a90fafd33ec18c0cbb/src/util/system.cpp#L291
https://github.com/bitcoin/bitcoin/blob/b6e34afe9735faf97d6be7a90fafd33ec18c0cbb/src/util/system.cpp#L439
https://github.com/bitcoin/bitcoin/blob/b6e34afe9735faf97d6be7a90fafd33ec18c0cbb/src/util/system.cpp#L827

I think this relates to "sections" of the configuration file. We haven't merged that yet, but we could someday.

Eirik0 replied:
I removed the `.` and `_`.

Eirik0 commented:
@zkbot r+

Eirik0 commented:
Passed on some workers but not others, looks like transient failures.
```
Running 310 test cases...
0%   10   20   30   40   50   60   70   80   90   100%
|----|----|----|----|----|----|----|----|----|----|
***************************[0;39;49
m************
********test/rpc_wallet_tests.cpp(294): error: in "rpc_wallet_tests/rpc_wallet": unexpected exception thrown by CallRPC("getblock 0")
test/rpc_wallet_tests.cpp(295): error: in "rpc_wallet_tests/rpc_wallet": unexpected exception thrown by CallRPC("getblock 0 0")
test/rpc_wallet_tests.cpp(296): error: in "rpc_wallet_tests/rpc_wallet": unexpected exception thrown by CallRPC("getblock 0 1")
test/rpc_wallet_tests.cpp(297): error: in "rpc_wallet_tests/rpc_wallet": unexpected exception thrown by CallRPC("getblock 0 2")
****
*** 4 failures are detected in the test module "Bitcoin Test Suite"
```
I am less familiar with the above transient failure, but those tests pass for me locally.

Eirik0 commented:
@zkbot retry

rex4539 commented:
Merging this broke the compilation on macOS https://github.com/zcash/zcash/issues/4164
